### ISSUE TITLE
Reimplement closures?

### ISSUE THREAD

**Issue by sumproxy** [crystal-lang/crystal#4866]
As far as I know closure in Rust are fast and don't incur much performance penalty, do you think it is possible to somehow import their implementation (better way to say: the ideas how to do that) to crystal?

This is submitted as an item for further discussion not as a bug or problem report.

**Comment by akzhan** [comment]
I'm newbee - what's wrong with current closures implementation?

It just works - https://play.crystal-lang.org/#/r/2klh

```crystal
def x
  a = 5
  b = ->() do
    p a
  end
end

x.call
```

**Comment by sumproxy** [comment]
I'm not an expert either, and this case is somewhat special, because you are returning a closure from a function, which requires a nightly rust's feature. It would look something like this in rust:

```rust
#![feature(conservative_impl_trait)]`
fn x() -> impl Fn(i32) {
    let a = 5;
    || println!("{}", a)
}

fn main() {
    for _ in 0..10_000 {
        x()();
    }
}
```

I was talking about more general cases and some kind of performance comparison (if crystal's closures are comparable performance-wise I'd fine with what there is).

**Comment by RX14** [comment]
Have you looked at blocks? I don't know how closures are used in rust but blocks actually have 0 overhead in crystal. That's why we can recommend using `5.times do |i|` instead of a while loop even in performance sensitive code. 

Also, if you're going to open an issue please do some more research. At the very least research *how* rust makes "fast closures" and provide some links.

**Comment by sumproxy** [comment]
Ok, sorry for the hassle.

**Comment by ysbaddaden** [comment]
This is a question for stackoverflow: [crystal-lang](https://stackoverflow.com/questions/tagged/crystal-lang).

**Comment by konovod** [comment]
Crystal blocks have some limitations.
5-minute research shows that Rust used to have "boxed closures" that are basically same as Crystal procs (two pointers, basically can't be inlined), but then introduced "unboxed closures" - an anonymous struct with fields for all captured vars and "call" operator implementation. Such closures can be inlined easier, but still can't replace boxed closures at all places. Basically they are somewhere between blocks and procs - they can e.g. be stored to variables and even returned from a function as @sumproxy shows.
Links:
https://github.com/rust-lang/rfcs/blob/master/text/0114-closures.md
tldr version: https://www.reddit.com/r/rust/comments/2lo6yt/closures_vs_unboxed_closures/

Imho having three types of "closures" is too much for a language so yes, fast closures aren't needed in Crystal if they can't replace `Proc`s in all uses.

**Comment by RX14** [comment]
So we have `yield` with zero overhead and `Proc` with some overhead (but really not that much except for some edge cases). That seems like quite a good tradeoff to me.
